File: code/WRSN-CEC2021-MMC-MFEA/src/algo/path/PIndividual.java
package algo.path;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import elements.Charger;
import elements.Sensor;
import problem.Configs;
import problem.ProblemManager;

public class PIndividual {
	private static long counter = 0;
	private long id;

	private ArrayList<Integer> path;
	private double[] chargingTime;

	private double[] gene;
	
	private int skillFactor;
	private double scalarFitness;
	private double fitness;
	private int[] factorialRank;

	public PIndividual() {
		id = ++counter;
		gene = new double[ProblemManager.USSD];
		this.factorialRank = new int[ProblemManager.getTaskNumber()];
	}

	public void randomInit() {
		for (int i = 0; i < gene.length; i++) {
			gene[i] = Configs.rand.nextDouble();
		}
	}

	@SuppressWarnings("unchecked")
	public void decoding(int taskIndex) {
		path = new ArrayList<Integer>();

		List<Allele> list = new ArrayList<Allele>();
		ArrayList<Sensor> sectors = ProblemManager.subNet[taskIndex];
		int n = sectors.size();

		for (int i = 0; i < n; i++) {
			Sensor s = sectors.get(i);
			if (s.getLifetime() > Configs.T) {
				gene[i] = 0.5 + 0.5 * Configs.rand.nextDouble();
			}
			if (gene[i] <= 0.5) {
				list.add(new Allele(gene[i], s.getId()));
			}
		}

		if (list.size() == 0) {
			list.add(new Allele(gene[0], sectors.get(0).getId()));
		}

		Collections.sort(list);
		for (Allele a : list) {
			path.add(a.getSensorId());
		}
	}

	public double calculateFitness() {
		this.decoding(skillFactor);
		double w[] = this.getPriority();

		int dead = 0;
		double totalLifetime = 0;
		double maxLifetime = 0;

		int n = path.size();
		double[] arriveTime = new double[n];
		boolean visited[] = new boolean[ProblemManager.maxSensorId + 1];
		chargingTime = new double[n];

		Charger ch = ProblemManager.chargers.get(skillFactor);
		double maxChargingTime = ch.getMaxChargingTime(path);

		// repairing
		arriveTime[0] = ProblemManager.distance[ProblemManager.serviceStation.getId()][path.get(0)] / ch.getSpeed();
		for (int i = 0; i < n; i++) {
			Sensor s = ProblemManager.getSensorById(path.get(i));
			chargingTime[i] = w[i] * maxChargingTime;
			visited[s.getId()] = true;

			double eRemain = s.getE0() - arriveTime[i] * s.getP();
			double ub = (Configs.S_EMAX - eRemain) / (ch.getU() - s.getP());
			if (eRemain < Configs.S_EMIN) {
				chargingTime[i] = 0;
			} else if (chargingTime[i] > ub || true) {
				chargingTime[i] = ub;
			}
			double eAfterT = s.getE0() - Configs.T * s.getP() + chargingTime[i] * ch.getU();

			if (eRemain < s.getEmin() - 1e-3 || eAfterT < s.getEmin() - 1e-3) {
				dead++;
			}

			double lifetime = (s.getE0() + chargingTime[i] * ch.getU() - s.getEmin()) / s.getP();
			totalLifetime += lifetime;
			maxLifetime = Math.max(lifetime, maxLifetime);

			if (i < n - 1) {
				arriveTime[i + 1] = arriveTime[i] + chargingTime[i]
						+ ProblemManager.distance[path.get(i)][path.get(i + 1)] / ch.getSpeed();
			}
		}

		for (Sensor s : ProblemManager.subNet[this.skillFactor]) {
			if (!visited[s.getId()]) {
				double lifetime = s.getLifetime();
				if (lifetime < Configs.T) {
					dead++;
				}

				totalLifetime += lifetime;
				maxLifetime = Math.max(lifetime, maxLifetime);
			}
		}

		double netSize = ProblemManager.subNet[this.skillFactor].size();
		double networkSurvivability = (netSize - dead) / (1.0 * netSize);
		double avgLifetime = totalLifetime / (1.0 * netSize);

		return networkSurvivability;
	}

	public double[] getPriority() {
		double[] priority = new double[path.size()];

		double sum = 0;
		for (int i = 0; i < path.size(); i++) {
			Sensor s = ProblemManager.getSensorById(path.get(i));
			priority[i] = s.getW();
			sum += priority[i];
		}

		for (int i = 0; i < path.size(); i++) {
			priority[i] /= sum;
		}

		return priority;
	}

	/**
	 * polynormial mutation
	 */
	public void mutation() {
		double u;
		for (int i = 0; i < gene.length; i++) {
			u = Configs.rand.nextDouble(); // generate random u in [0, 1]
			if (u <= 0.5) {
				double deltaL = Math.pow(2 * u, 1.0 / (Configs.NM + 1)) - 1;
				double p = this.getGene(i) * (1.0 + deltaL);
				this.setGene(i, Math.max(0, Math.min(1, p)));
			} else {
				double deltaR = 1.0 - Math.pow(2 * (1 - u), 1.0 / (Configs.NM + 1));
				double p = this.getGene(i) + deltaR * (1 - this.getGene(i));
				this.setGene(i, Math.max(0, Math.min(1, p)));
			}
		}
	}

	public double[] getGene() {
		return gene;
	}

	public void setGene(double[] gene) {
		this.gene = gene;
	}

	public void setGene(int index, double value) {
		gene[index] = value;
	}

	public double getGene(int index) {
		return gene[index];
	}

	public ArrayList<Integer> getPath() {
		return path;
	}

	public void setPath(ArrayList<Integer> path) {
		this.path = path;
	}

	public int getSkillFactor() {
		return skillFactor;
	}

	public void setSkillFactor(int skillFactor) {
		this.skillFactor = skillFactor;
	}

	public double getScalarFitness() {
		return scalarFitness;
	}

	public void setScalarFitness(double scalarFitness) {
		this.scalarFitness = scalarFitness;
	}

	public double getFitness(int task) {
		if (task == this.skillFactor) {
			return fitness;
		} else {
			return -Double.MAX_VALUE;
		}
	}

	public void setFitness(double value) {
		fitness = value;
	}

	public int[] getFactorialRank() {
		return factorialRank;
	}

	public int getFactorialRank(int task) {
		return factorialRank[task];
	}

	public void setFactorialRank(int[] factorialRank) {
		this.factorialRank = factorialRank;
	}

	public void setFactorialRank(int task, int factorialRank) {
		this.factorialRank[task] = factorialRank;
	}

	public long getId() {
		return id;
	}

	public double[] getChargingTime() {
		return chargingTime;
	}

	public void setChargingTime(double[] chargingTime) {
		this.chargingTime = chargingTime;
	}

	public double getChargingTime(int index) {
		return this.chargingTime[index];
	}

	public PIndividual clone() {
		PIndividual indiv = new PIndividual();
		indiv.setGene(this.getGene().clone());
		return indiv;
	}

}

File: algo/path/PIndividual.py
import sys
from problem.ProblemManager import ProblemManager
from problem.Configs import Configs
from algo.path.Allele import Allele
from elements.Charger import Charger
class PIndividual:
    counter: float
    """n/a"""
    counter = 0

    id: float
    """n/a"""

    path: 'list[int]'
    """n/a"""

    chargingTime: 'list[float]'
    """"""

    gene: 'list[float]'
    """n/a"""

    skillFactor: int

    scalarFitness: float

    fitness: float
    fitness = 0.0

    factorialRank: 'list[int]'

    def __init__(self) -> None:
        self.counter = self.counter + 1
        self.id = self.counter

        self.gene = []
        for i in range(ProblemManager.USSD):
            self.gene.append(Configs.rand.uniform(0,1))

        self.factorialRank = []
        for i in range(ProblemManager.getTaskNumber()):
            self.factorialRank.append(-1)
        
    def randomInit(self):
        for i in range(ProblemManager.USSD):
            self.gene[i] = Configs.rand.uniform(0,1)

    def decoding(self, taskIndex: int):
        self.path = []

        list: 'list[Allele]'
        list = []

        #TODO Subnet may be empty. 
        #not a real copy, just copy list's reference
        sectors = ProblemManager.subNet[taskIndex]

        for index, sector in enumerate(sectors):
            # if sector.getLifetime() > Configs.T:
            #     self.gene[index] = 0.5 + 0.5 * Configs.rand.uniform(0,1)
            if self.gene[index] <= 0.5:
                list.append(Allele(self.gene[index],sector.id))

        if len(list) == 0:
            list.append(Allele(self.gene[0],sectors[0].id))
        
        #bug prone!!
        list.sort()

        for a in list:
            self.path.append(a.sensorId)
        pass
    def calculateFitnessWithPath(self) -> float:
        w: list[float]
        w = self.getPriority()

        dead = 0
        totalLifetime = 0
        maxLifetime = 0
        totalAfterT=0
        n = len(self.path)
        arriveTime: 'list[float]'
        arriveTime = []
        for i in range(n):
            arriveTime.append(0.0)

        visited: 'list[bool]'
        visited = []
        for i in range(ProblemManager.maxSensorId+1):
            visited.append(False)


        self.chargingTime = []
        for i in range(n):
            self.chargingTime.append(0)

        ch: Charger
        ch = ProblemManager.chargers[self.skillFactor]

        maxCharingTime: float
        maxCharingTime = ch.getMaxChargingTime(self.path)

        #repairing
        arriveTime[0] = ProblemManager.distance[ProblemManager.serviceStation.id][self.path[0]] /ch.speed

        totalTimeRatio = 0.0

        for i in range(n):
            s = ProblemManager.getSensorById(self.path[i])
            self.chargingTime[i] = w[i] * maxCharingTime
            visited[s.id] = True

            eRemain: float
            ub: float
            eRemain = s.E0 - arriveTime[i] * s.p
            # ub = (Configs.S_EMAX - eRemain) / (ch.U - s.p)
            if eRemain < Configs.S_EMIN:
                self.chargingTime[i] = 0
            # elif (self.chargingTime[i] > ub):
            #     self.chargingTime[i] = ub
            eAfterT: float
            eAfterT = s.E0 - Configs.T * s.p + self.chargingTime[i] * ch.U
            if ( (eRemain < s.Emin - 1e-3) or (eAfterT < s.Emin - 1e-3) ) :
                dead = dead + 1
			
            lifetime: float
            lifetime = (s.E0 + self.chargingTime[i] * ch.U - s.Emin) / s.p

            ######
            ratio= (eAfterT/s.p) / Configs.T
            if ratio>1:
                ratio=1
            totalTimeRatio+=ratio 
            totalLifetime += lifetime
            maxLifetime = max(lifetime, maxLifetime)
            totalAfterT+=eAfterT

            if (i < n - 1):
                arriveTime[i + 1] = arriveTime[i] + self.chargingTime[i] \
                        + ProblemManager.distance[self.path[i]][self.path[i+1]] / ch.speed
            
        for s in ProblemManager.subNet[self.skillFactor]:
            if (not visited[s.id]):
                lifetime = s.getLifetime()
                if (lifetime < Configs.T):
                    dead = dead + 1

                totalLifetime += lifetime
                maxLifetime = max(lifetime, maxLifetime)
                eAfterT = s.E0 - Configs.T * s.p
                totalAfterT+=eAfterT



        netSize = len(ProblemManager.subNet[self.skillFactor])
        networkSurvivability = (netSize - dead) / (1.0 * netSize)
        energyRatioAfterT=totalAfterT/(netSize*Configs.S_EMAX)
        totalTimeRatio=totalTimeRatio/netSize
        #unused
        #avgLifetime = totalLifetime / (1.0 * netSize)
        # return networkSurvivability
        # print("check",networkSurvivability," va ",totalTimeRatio)


        return networkSurvivability*Configs.networkSurvivabilityFitness+totalTimeRatio*Configs.totalTimeRatioFitness

        
        # return networkSurvivability*0.8+0.2*energyRatioAfterT   

    def calculateFitness(self) -> float:
        """How good the individual is"""
        self.decoding(self.skillFactor) #from gene -> get path
        return self.calculateFitnessWithPath()

    def getPriority(self) -> 'list[float]':
        sum = 0.0
        priority: 'list[float]'
        priority = []
        for i in range (len(self.path)):
            s = ProblemManager.getSensorById(self.path[i])
            priority.append(s.getW())
            sum += priority[i]

        for i in range (len(self.path)):
            priority[i] /= sum

        return priority

    def mutation(self) -> None:
        u: float
        for i in range(len(self.gene)):
            u = Configs.rand.uniform(0,1)
            if u <= 0.5:
                deltaL = pow(2 * u, 1.0 / (Configs.NM + 1)) - 1
                p = self.gene[i] * (1.0 + deltaL)
                # self.gene[i]= 0.5+deltaL*(0.5-self.gene[i])
                self.gene[i] = max(0,min(1,p))
                # print("u<0.5",self.gene[i],p)
            else:
                deltaR = 1.0 - pow(2 * (1 - u), 1.0 / (Configs.NM + 1))
                p = self.gene[i] + deltaR * (1 - self.gene[i])
                # self.gene[i]= 0.5+deltaR*(self.gene[i]-0.5)
                self.gene[i] = max(0,min(1,p))
                # print("u>0.5",self.gene[i],p)
    
            
    def getFitness(self, task: int) -> float:
        if task == self.skillFactor:
            return self.fitness
        else:
            return -sys.float_info.max

    def clone(self) -> 'PIndividual':
        indiv = PIndividual()
        indiv.gene = self.gene.copy()
        indiv.skillFactor=self.skillFactor
        indiv.path=self.path.copy()
        indiv.fitness=self.fitness
        return indiv

    def duplicate(self) -> 'PIndividual':
        """Everything from the old object will be kept

        This is a workaround to bypass the reference problem"""
        indiv = PIndividual()
        indiv.counter = self.counter
        indiv.id = self.id
        indiv.path = self.path.copy()
        indiv.chargingTime = self.chargingTime.copy()
        indiv.gene = self.gene.copy()
        indiv.skillFactor = self.skillFactor
        indiv.scalarFitness = self.scalarFitness
        indiv.fitness = self.fitness
        indiv.factorialRank = self.factorialRank.copy()
        return indiv

        # 2 opt
    def tsp_2_opt(self)->'PIndividual':
        if len(self.path)<15:
            return self
        sectors = ProblemManager.subNet[self.skillFactor]
        sectorsDict= {sectors[i].id: i for i in range(0, len(sectors))}
        
        improved = 0
        best_found_individual = self.clone()
        loopCounter = 0
        while improved<10:
            loopCounter += 1
            if loopCounter > 5000: #prevent dead loop
                break
            #print(f"2 opt improved {improved}, loop counter {loopCounter}")
            i =Configs.rand.randint(1,len(best_found_individual.path) - 2)
            j= Configs.rand.randint(i+1,len(best_found_individual.path) - 1)
            
            new_individual = self.swap_2opt(best_found_individual, i, j,sectorsDict)
            new_individual.fitness=new_individual.calculateFitnessWithPath()
            if new_individual.fitness < best_found_individual.fitness:
                best_found_individual = new_individual
                improved += 1
                loopCounter = 0
          
        return best_found_individual

    def swap_2opt(self,best_found_individual:'PIndividual',i:int,j:int,sectorsDict:'dict')->'PIndividual':
        
        new_individual=best_found_individual.clone()
        new_individual.path = best_found_individual.path[0:i]
        new_individual.path.extend(reversed(best_found_individual.path[i:j + 1]))
        new_individual.path.extend(best_found_individual.path[j + 1:])
   
        for t in range(i,j+1):
            a=sectorsDict[best_found_individual.path[t]]
            b=sectorsDict[new_individual.path[t]]
            new_individual.gene[b]=best_found_individual.gene[a]
        return new_individual

# 3 opt
    def tsp_3_opt(self)->'PIndividual':
        if len(self.path)<15:
            return self
        moves_cost:list[PIndividual]
        sectors = ProblemManager.subNet[self.skillFactor]
        sectorsDict= {sectors[i].id: i for i in range(0, len(sectors))}
        loopCounter = 0
        improved = 0
        best_found_individual = self.clone()
        best_found_individual.fitness=best_found_individual.fitness
        while improved<3:
            loopCounter += 1
            if loopCounter > 5000: #prevent dead loop
                break
            #print(f"3 opt improved {improved}, loop counter {loopCounter}")
            k= Configs.rand.randint(2,len(best_found_individual.path) - 2)
            i =Configs.rand.randint(1,k)
            j= Configs.rand.randint(k+1,len(best_found_individual.path) - 1)

        
            moves_cost = self.get_solution_cost_change(best_found_individual, i, j, k)
            # we need the minimum value of substraction of old route - new route
            best_return_3_opt = min(moves_cost, key=lambda item: item.fitness)
            # print(f'compare {best_return_3_opt.fitness} va {best_found_individual.fitness}')
            if best_return_3_opt.fitness < best_found_individual.fitness:
                for t in range(len(best_found_individual.path)):
                    a=sectorsDict[best_found_individual.path[t]]
                    b=sectorsDict[best_return_3_opt.path[t]]
                    best_return_3_opt.gene[b]=best_found_individual.gene[a]
                # print("after opt decreas",best_found_individual.fitness-best_return_3_opt.fitness)
                best_found_individual = best_return_3_opt
                improved += 1
                loopCounter = 0
        # just to start with the same node -> we will need to cycle the results.
        # print(f'{len(best_found_individual.path)}:{improved}:{loopCounter}')
        return best_found_individual

    def get_solution_cost_change(self,best_found_individual:'PIndividual', i:int, j:int, k:int)->'list[PIndividual]':
        number_Of_3_OPT_case=8
        new_individuals:list[PIndividual]
        new_individuals=[]
        for i in range(number_Of_3_OPT_case):
            new_individual=best_found_individual.clone()
            new_individuals.append(new_individual)

        new_individuals[0].path =  new_individuals[0].path[:i + 1] +  new_individuals[0].path[i+1:k + 1] + new_individuals[0].path[k+1: j + 1] + new_individuals[0].path[j+1: ]
        new_individuals[1].path =  new_individuals[1].path[:i + 1] +  new_individuals[1].path[i+1:k + 1] + new_individuals[1].path[j: k: -1] + new_individuals[1].path[j+1: ]
        new_individuals[2].path =  new_individuals[2].path[:i + 1] +  new_individuals[2].path[k:i: -1] +  new_individuals[2].path[k+1: j + 1] +  new_individuals[2].path[j+1: ]
        new_individuals[3].path =  new_individuals[3].path[:i + 1] +  new_individuals[3].path[k:i: -1] +  new_individuals[3].path[j: k: -1] +  new_individuals[3].path[j+1: ]
        new_individuals[4].path =  new_individuals[4].path[:i + 1] +  new_individuals[4].path[k+1: j + 1] +  new_individuals[4].path[i+1:k + 1] +  new_individuals[4].path[j+1: ]
        new_individuals[5].path =  new_individuals[5].path[:i + 1] +  new_individuals[5].path[k+1: j + 1] +  new_individuals[5].path[k:i: -1] +  new_individuals[5].path[j+1: ]
        new_individuals[6].path =  new_individuals[6].path[:i + 1] +  new_individuals[6].path[j: k: -1] +  new_individuals[6].path[i+1:k + 1] +  new_individuals[6].path[j+1: ]
        new_individuals[7].path =  new_individuals[7].path[:i + 1] +  new_individuals[7].path[j: k: -1] +  new_individuals[7].path[k:i: -1] +  new_individuals[7].path[j+1: ]
        
        for i in range(number_Of_3_OPT_case):
            new_individuals[i].fitness=new_individuals[i].calculateFitnessWithPath()
        return new_individuals
            
    def __eq__(self, other: 'PIndividual'):
        if self.fitness == other.fitness:
            return True
        else:
            return False

    def __lt__(self, other: 'PIndividual'):
        if self.fitness > other.fitness:
            return True
        else:
            return False

    def __gt__(self, other: 'PIndividual'):
        if self.fitness < other.fitness:
            return True
        else:
            return False
    
    












